### ISSUE TITLE
Socket closed errors appear very frequently

### ISSUE THREAD

**Issue by shahamit** [microsoftgraph/msgraph-sdk-java#553]
We are using 2.3.0 version of the microsoft graph library and okhttp library version is 3.12.1. We see the below socket closed exception very frequently. This error is not specific to a particular graph api. In our application we call graph api's on the teams and drive resource.

We tried disabling the implicit retries from okhttp library by following the guidelines from [this link](https://docs.microsoft.com/en-us/graph/sdks/customize-client?tabs=java) but that didn't help. In fact the "connection reset by peer" error reappeared on doing so.

This issue is newly created after following the discussion on an [earlier issue](https://github.com/microsoftgraph/msgraph-sdk-java/issues/313).

com.microsoft.graph.core.ClientException: Error during http request at com.microsoft.graph.http.CoreHttpProvider.sendRequestInternal(CoreHttpProvider.java:458) at com.microsoft.graph.http.CoreHttpProvider.send(CoreHttpProvider.java:207) at com.microsoft.graph.http.CoreHttpProvider.send(CoreHttpProvider.java:187) at com.microsoft.graph.http.BaseRequest.send(BaseRequest.java:345) at com.microsoft.graph.http.CustomRequest.get(CustomRequest.java:43) at com.optymyze.officeintegration.alternateteams.GraphApiBatchService.getOrRetryGraphApiRequest(GraphApiBatchService.java:130) ... 15 more Caused by: java.net.SocketTimeoutException: timeout at okio.Okio$4.newTimeoutException(Okio.java:232) at okio.AsyncTimeout.exit(AsyncTimeout.java:285) at okio.AsyncTimeout$2.read(AsyncTimeout.java:241) at okio.RealBufferedSource.indexOf(RealBufferedSource.java:354) at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:226) at okhttp3.internal.http1.Http1Codec.readHeaderLine(Http1Codec.java:215) at okhttp3.internal.http1.Http1Codec.readResponseHeaders(Http1Codec.java:189) at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:88) at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45) at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93) at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93) at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:126) at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) at com.microsoft.graph.httpcore.TelemetryHandler.intercept(TelemetryHandler.java:37) at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) at com.microsoft.graph.httpcore.RedirectHandler.intercept(RedirectHandler.java:123) at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) at com.microsoft.graph.httpcore.RetryHandler.intercept(RetryHandler.java:140) at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) at com.microsoft.graph.httpcore.AuthenticationHandler.intercept(AuthenticationHandler.java:31) at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:254) at okhttp3.RealCall.execute(RealCall.java:92) at com.microsoft.graph.http.CoreHttpProvider.sendRequestInternal(CoreHttpProvider.java:385) ... 45 more Caused by: javax.net.ssl.SSLException: Socket closed at java.base/sun.security.ssl.Alert.createSSLException(Alert.java:127) at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:320) at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:263) at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:258) at java.base/sun.security.ssl.SSLSocketImpl.handleException(SSLSocketImpl.java:1314) at java.base/sun.security.ssl.SSLSocketImpl$AppInputStream.read(SSLSocketImpl.java:839) at okio.Okio$2.read(Okio.java:140) at okio.AsyncTimeout$2.read(AsyncTimeout.java:237) ... 77 more Caused by: java.net.SocketException: Socket closed at java.base/java.net.SocketInputStream.read(SocketInputStream.java:183) at java.base/java.net.SocketInputStream.read(SocketInputStream.java:140) at java.base/sun.security.ssl.SSLSocketInputRecord.read(SSLSocketInputRecord.java:448) at java.base/sun.security.ssl.SSLSocketInputRecord.bytesInCompletePacket(SSLSocketInputRecord.java:68) at java.base/sun.security.ssl.SSLSocketImpl.readApplicationRecord(SSLSocketImpl.java:1104) at java.base/sun.security.ssl.SSLSocketImpl$AppInputStream.read(SSLSocketImpl.java:823) ... 79 more

 
[AB#7125](https://microsoftgraph.visualstudio.com/0985d294-5762-4bc2-a565-161ef349ca3e/_workitems/edit/7125)

**Comment by shahamit** [comment]
@baywet - Continuing the discussion on [from previous issue](https://github.com/microsoftgraph/msgraph-sdk-java/issues/313#issuecomment-718712838)

> hostname: if you use upload sessions the hostname will be something like https://sn3302.up.1drv.com/up/fe6987415ace7X4e1eF866337 (for OneDrive, each service has their own upload URLs). The access token you get is only valid for a specific audience (graph.microsoft.com) and not valid for any other domain. Sending it along with upload requests will cause an access denied. The reason why you don't need an access token for those URLs is because they are either short-lived or they contain the access token as a query string parameter. (one valid for that audience)

We do use [this ](https://docs.microsoft.com/en-us/graph/api/driveitem-createuploadsession?view=graph-rest-1.0)api for uploading files on sharepoint but we haven't seen any access denied errors. The code does use the same authenticator due to which the access token does get passed as a request header. Could the difference be that we are using sharepoint (and not one drive)?

> What did your investigations disabling the retry resulted in?

Disabling the okhttp retries didn't help. In fact we again started seeing "connection reset by peer" error that we had seen [earlier](https://github.com/microsoftgraph/msgraph-sdk-java/issues/313#issuecomment-694078212), along with socket closed errors. 

> Are all your customer instances behind the same firewall?

Yes. All customers deployed in one AWS region are behind the same firewall and having a separate firewall is way to costly (may not be worthwhile too). As per the deployment team this is the standard deployment followed generally by everyone. 

If these errors are appearing due to api rate limits then I am not sure what we can do from our side since we don't know how many parallel requests could cause the server to drop api requests. 

What are your suggestions/recommendations? 
I am curious, how do other customers manage it. Could you please share some insights?

**Comment by baywet** [comment]
**hostname:**
yes, APIs for different services might behave differently unfortunately. It might be that this API in SharePoint ignores the access token for upload URLs when the same one for Exchange does not. If it's not broken don't fix it but expect to see issues with your current implementation if you start uploading email attachments (or in other cases).

**connections dropping:** do you mind sharing volume information please? I'm particularly interested in how many parallels connections you have from the same IP address (average and max) and how many parallels HTTP requests you're making for the same IP address (average and max).
Solutions used by our customers vary, and we yet have to diagnose why you're getting connections dropped at this point, but here are a few pointers:
- caching to limit read requests
- using multiple public IPs
- some kind of pool/proxy/circuit breaker mechanism to control/limit the number of outgoing connections to Microsoft Graph
- features shedding (i.e. delay some of the calls that can wait to later times when the application is less busy)

**Comment by shahamit** [comment]
> do you mind sharing volume information please? I'm particularly interested in how many parallels connections you have from the same IP address (average and max) and how many parallels HTTP requests you're making for the same IP address (average and max).

This information is a bit hard to capture and we are working to find a monitoring solution for this. It would take some time.

> Can you set the client-request-id (guid) header on the requests and log it please? (more-reading).

Yes, we will log the request id and share them as soon as possible.

**Comment by shahamit** [comment]
@baywet - We are making the code changes to log the client-request-id but we also thought we would log the response headers when errors appear. It seems only the `GraphServiceException` class has methods to log the response headers but not `ClientException` class and this (`ClientException`) is the exception that we get when socket gets closed. Can you guide us how do log response headers for `ClientException`?

**Comment by baywet** [comment]
GraphServiceExceptions represent the errors returned by the service itself, it relies on an HTTP response to parse the error and get the response headers. ClientException represent other types of errors (happening in the application, client library, network between the app and Graph...).
From the stacktrace you provided, it looks like the connection is closed before the SDK even receives a response from the service, which means we won't be able to get response headers. It we were getting the response to begin with, you'd get a GraphServiceException instead of what you're currently receiving.
I think you should test on whether the exception you're getting is a graph service exception and if it is, then log the response headers.

**Comment by shahamit** [comment]
> From the stacktrace you provided, it looks like the connection is closed before the SDK even receives a response from the service, which means we won't be able to get response headers.

Does that indicate that the requests are getting dropped at azure or aws?

> I think you should test on whether the exception you're getting is a graph service exception and if it is, then log the response headers.

All the exceptions that we are getting are instances of `ClientException`. We have added code to log the response headers in the catch block of `GraphServiceException` but that's just a safeguard. Let's see if we can troubleshoot these socket closed errors by logging the client-request-id and timestamps. 

About logging client-request-id for every request, we realized that our code base has around 30 graph api calls spread across multiple source files. We could write util methods that generate, log and pass the request id when calling the `buildRequest()` method but we thought to check with you to see if the graph sdk can help out here to make this code central and easy to read? Some sort of interceptor?

**Comment by ravisamani** [comment]
Also after we started passing client-request-id to request, I was expecting response header to have same one.
but it's different:
e.g. 
request header:     client-request-id : client-request-id-241ikwmfvdlswdl28y
response header:  client-request-id : 3870774f-b7b2-42ef-a971-9126a77de96d

So @baywet which one is useful for you?
any reason why response still generates new one?

cc: @shahamit

**Comment by baywet** [comment]
**systematically inject the header:** yes, you can use an [okhttp interceptor](https://github.com/microsoftgraph/msgraph-sdk-java-core/blob/dev/src/main/java/com/microsoft/graph/httpcore/RedirectHandler.java), [here is how you customize the client (to inject it)](https://docs.microsoft.com/en-us/graph/sdks/customize-client?tabs=java).

**response header being different:** can you try injecting a guid instead of that custom value please?

**Comment by ravisamani** [comment]
Thanks @baywet  GUID works 👍 

About customizing client:
>   [here is how you customize the client (to inject it)](https://docs.microsoft.com/en-us/graph/sdks/customize-client?tabs=java).

Snippet uses two different authProviders: 

1. com.microsoft.graph.authentication.IAuthenticationProvider
2. com.microsoft.graph.httpcore.ICoreAuthenticationProvider

@baywet How to build GraphServiceClient without using  @Deprecated one - com.microsoft.graph.authentication.IAuthenticationProvider ?

**Comment by baywet** [comment]
Today you can't. I added the deprecated information a few versions ago to signal we are going to move it soon. I'm not sure it was a good idea anymore, see #545 
I'm working on cleaning up all the auth story with [version 3](https://github.com/microsoftgraph/msgraph-sdk-java/milestone/4)

**Comment by shahamit** [comment]
@baywet - we were in a festive break since last week. We are almost done with the code changes to log the client-request-id and should be able to soon share them with you.

**Comment by baywet** [comment]
Happy Diwali!

**Comment by bora-fsd** [comment]
@baywet any idea approximately when this will be released? We're close to rolling out our own custom solution (using try/catch and retry connection a few times) but if there's a better solution in the SDK, we'd rather wait for it.

**Comment by baywet** [comment]
@bora-fsd at this point in time we're not working on a fix in the SDK, we're still trying to identify where is the issue coming from.

I'm not sure how much have you been reading/following on this issue but we already [enabled connection retry on drop](https://github.com/microsoftgraph/msgraph-sdk-java/issues/313), [disabled HTTP/2 by default](https://github.com/microsoftgraph/msgraph-sdk-java/issues/554) as the service doesn't support it today and the http library we're using has a known issue in that case with some runtimes, and [fixed leaky connection management](https://github.com/microsoftgraph/msgraph-sdk-java/issues/284).
So if you're not already on 2.3.2 of the library and 1.0.5 of core, I'd encourage you to give it a try.

The only other thing that'd impact behavior at this point and which is under work would be the update from okhttp 3 to 4 that's upcoming with v3 of the library, but we're still a month out at least.
I hope these additional details help set the context.

**Comment by ravisamani** [comment]
Adding details for one of the Socket Closed Error occurred on **November 26, 00:09:19**

We checked total calls made from our servers to graph for duration of that minute: Count is **187** calls.

Please check bar chart below: 
![187 Graph Calls in 1 minute](https://user-images.githubusercontent.com/7370268/100453877-c8d35500-30e1-11eb-905a-86401c3053af.png)

Each bar - represent : 
- one second on X axis
- count of Graph API calls on Y Axis ( Max count per second is 7 )





> @baywet - we were in a festive break since last week. We are almost done with the code changes to log the client-request-id and should be able to soon share them with you.

**Comment by baywet** [comment]
do you have the client request id of the request that received a socket closed? I'd like to see in the logs whether it actually got to Microsoft Graph or not.

**Comment by ravisamani** [comment]
Sure,

Failed URL: https://graph.microsoft.com/beta/users/{user-id}/joinedTeams?%24select=id%2CdisplayName%2Cdescription

client-request-id: **ae121188-86df-4070-9cc3-e84a945f3745**

**Comment by shahamit** [comment]
We analyzed another regression run that had the below stats

There were 37456 graph api requests made on 24th Nov test run. Here's how the bar graph is for it

![image](https://user-images.githubusercontent.com/1281926/100605711-530ff900-332e-11eb-922e-6ce0183b4075.png)

There were around 6 socket closed errors during this time as below
		a. Client request id - 9d999c8b-5156-4039-bb01-5a8d045d3243. Timestamp (GMT) - November 24th 2020, 19:01:04.506
		b. Client request id - 7a709927-7f46-481a-bd04-dd8e57bf443a. Timestamp (GMT) - November 24th 2020, 19:11:32.989
		c. Client request id - 236057ad-346c-4a3d-9766-dc09e7ad3505. Timestamp (GMT) - November 24th 2020, 19:12:57.884
		d. Client request id - 34ffbbbd-038a-4d1a-a947-5857690da1e6. Timestamp (GMT) - November 24th 2020, 19:14:44.749
		e. Client request id - 94b6baef-2df9-487d-b1c4-9b683fbd7749. Timestamp (GMT) - November 24th 2020, 19:27:28.751
                f. Client request id - 85e8b247-fa35-42ed-9827-95ba419860c2. Timestamp (GMT) - November 24th 2020, 19:44:43.217

**Comment by baywet** [comment]
Thanks for providing this additional information. Here are the results:
a. not found in the logs
b. 200 OK
c. not found in the logs
d. not found in the logs 
e. not found in the logs
f. 200 OK

I have a discussion scheduled on late Wednesday with a backend engineer to try to understand better what's going on here. While we wait, it'd be worth investigating for [port exhaustion ](https://docs.microsoft.com/en-us/windows/client-management/troubleshoot-tcpip-port-exhaust) and [try to tweak the connection pool](https://github.com/square/okhttp/issues/4354).

Here is my thinking, since you're getting socket closed even on 200, and some other requests don't even make it to the service, it looks like your "application" gets "hot" and starts to randomly drop connections, either before the request has made it to the service or before the response has been received. This, in my opinion, happens for one of three reasons:

- Your application instantiates too the GraphServiceClient/OkHttpClient too often. This leads to [creating a different pools](https://square.github.io/okhttp/concurrency/#connection-pool) instead of reusing existing pools, which leads to creating more connections.
- Connection pool is [not properly configured](https://github.com/square/okhttp/issues/4354), leading to OkHttpClient to evict some connections too frequently and port exhaustion.
- Something is leaky in this library or in your application, keeping connections open and leading to exhaustion.

Note: there could be other network/OS/JVM considerations here, but I think the hypothesis I listed will be easier to test for.

Could you:

- Monitor network activity, specifically idle ports for a single or few PIDs, which would be signs of port exhaustion.
- Try to increase the keep alive duration, and see if it improves the issue.
- Add the -verbose:gc argument to the JVM and monitor for signs of leaky memory management.
- Share more details on how often a GraphServiceClient/OkHttpClient gets instantiated throughout the application? Especially since you're in a tomcat application which, as far as I understand, creates threads to handle incoming requests.

**Comment by shahamit** [comment]
Thanks for the analysis. Possibly there could also be two different reasons why we get socket closed errors. The ones that show up as 200 OK in the graph api server logs could be because of the client side issues as mentioned above and the ones that don't show up in the logs could also be because they get blocked by the azure service end too right? Did you check those logs?

From the numbers it seems graph api rate limits are not that high too. I could see around 3k api calls within 5 mins. We are also reviewing our tests to understand the api call pattern.

**Comment by SrujanDacharaju** [comment]
@baywet do you have timeline for build when this error will be fixed?

**Comment by baywet** [comment]
Just to be clear I'll re-iterate: at this point we're not planning any change to the SDK for that specific issue. The service team is still investigating on their side to get a better understanding of the problem.
I'm also waiting on feedback from @shahamit  and @ravisamani  regarding the investigations listed in [this comment](https://github.com/microsoftgraph/msgraph-sdk-java/issues/553#issuecomment-735830313), especially the one about tweaking the connection pool with longer idle max times and max idle connections so I can share that information with the service team.

**Comment by SrujanDacharaju** [comment]
Thank you @baywet . Meanwhile , do you suggest any alternative?

**Comment by baywet** [comment]
Try out the investigations I list in the comment I linked and see if it improves the situation (and please do report the results here).
A global wrapper/retry handler to catch these network drops and retry could also help as a temporary workaround, but Okhttp already does a lot on that front as explained above.

**Comment by megicivovic** [comment]
Hello @baywet , this is my use case with Microsoft Graph API. I think the issue here is having to instantiate the graph client every time the access token changes, which results in creating too many http connections. Can you please explain why is it not possible to set the token on the GraphServiceClient instance, but rather have to create a new builder instance every time it changes? Also, can you tell me if this is the correct usage of the client, or should I somehow cache the http client as suggested in this [issue](https://stackoverflow.com/questions/52254481/what-is-the-recommendation-to-reuse-graphserviceclient-in-microsoft-graph-net-s)? Thanks in advance!

`  public IGraphServiceClient getClient(@Nonnull String accessToken) {
    final SimpleAuthProvider authProvider = new SimpleAuthProvider(accessToken);

    final IGraphServiceClient graphServiceClient =
        GraphServiceClient.builder()
            .authenticationProvider(authProvider)
            .logger(new DefaultLogger())
            .buildClient();
    graphServiceClient.setServiceRoot(providerConfig.getUrlBeta());
    return graphServiceClient;
  }`

**Comment by baywet** [comment]
@megicivovic Thanks for going over this issue and investigating on your end.
Where does this `SimpleAuthProvider` come from? Why does it accept the auth token instead of a callback? This is generally a bad pattern to set the token once in the provider and not have code running an evaluation on every call the client might execute.

The reason why we don't allow to set the auth token on the client right away is because that token might be expired, underprivileged, not for the right resource... and the auth middleware needs to call the auth provider so extraction from cache/renewal/grant/... can be done.

Depending on the application type (backend, native,...) we do encourage to reuse the GraphClient instead of re-instantiating it for the reasons outlined in the stack overflow post you linked.
An alternative could be for your application to hold a static instance of OkHttpClient (assuming you're on 2.X) and pass it to the GraphClient builder. This way the Graph client would be new but the OkHttpClient (which holds the http connection pool) would be static, see [this documentation](https://docs.microsoft.com/en-us/graph/sdks/customize-client?tabs=java). This is far from the better option as the GraphClient also holds service executors in 2.X and might hold some client side caching in the future, it'd be better to reuse the client instead and have an authentication provider accept a callback which could evaluate the situation.

**Comment by megicivovic** [comment]
@baywet the` SimpleAuthProvider` cam from your [SDK docs](https://docs.microsoft.com/en-us/graph/tutorials/java?tutorial-step=4#adding-calendar-and-contacts-apis).

I have refactored my code with your suggestions , so it now looks like this
`  public IGraphServiceClient getClient(@Nonnull final String accessToken) {

    final OkHttpClient httpClient =
        HttpClients.createDefault(
                request ->
                    request
                        .newBuilder()
                        .addHeader(Keys.AUTHORIZATION, Keys.BEARER.concat(accessToken))
                        .build())
            .newBuilder()
            .followSslRedirects(false) // sample configuration to apply to client
            .build();

    final IAuthenticationProvider iAuthenticationProvider =
        request -> request.addHeader(Keys.AUTHORIZATION, Keys.BEARER.concat(accessToken));

    final IHttpProvider httpProvider =
        DefaultClientConfig.createWithAuthenticationProvider(iAuthenticationProvider)
            .getHttpProvider(httpClient);

    final IGraphServiceClient graphServiceClient =
        GraphServiceClient.builder()
            .authenticationProvider(iAuthenticationProvider)
            .httpProvider(httpProvider)
            .logger(new DefaultLogger())
            .buildClient();

    graphServiceClient.setServiceRoot(providerConfig.getUrlBeta());
    return graphServiceClient;
  }`
The only question I have is - even if aI had a static instance of OkHttpClient, I still have to use 
`HttpClients.createDefault(
                request ->
                    request
                        .newBuilder()
                        .addHeader(Keys.AUTHORIZATION, Keys.BEARER.concat(accessToken))
                        .build())`
 to pass the new access token. Does this seem reasonable to you? Thank you for all of your help!

**Comment by baywet** [comment]
oh I get it! You are stuck because of the API that's currently provided, in any case your application needs an authentication provider which in turns depends on the access token. What's needed is for you to be able to set the authentication provider on the client.

In the meantime you could imagine doing something like that. However this is likely not to be thread safe and might need some locking.

```Java
private static IGraphServiceClient staticClient =  GraphServiceClient.builder()
            .authenticationProvider(new DynamicAccessTokenProvider())
            .buildClient();

public IGraphServiceClient getClient(@Nonnull final String accessToken) {
    final IGraphServiceClient graphServiceClient = staticClient;
    ((DynamicAccessTokenProvider)graphServiceClient.getAuthenticiationProvider()).tokenSupplier = () -> accessToken;
    graphServiceClient.setServiceRoot(providerConfig.getUrlBeta());
    return graphServiceClient;
  }

public class DynamicAccessTokenProvider implements IAuthenticationProvider, ICoreAuthenticationProvider {
  // the implementation is most likely not right but I'm keeping it short for simplicity and to demonstrate the point
  public Supplier<String> tokenSupplier;
  public IHttpRequest authenticateRequest(IHttpRequest request) {
    if(tokenSupplier != null)
      request.addHeader(Keys.AUTHORIZATION, Keys.BEARER.concat(tokenSupplier.get()));
  }
}
```

**Comment by megicivovic** [comment]
@baywet , this is exactly what I needed. Thank you very much!

**Comment by baywet** [comment]
Thanks for the follow up. After a few days this newer version of the code has been deployed to production, can you report on the impact on connections management and the improvements please? This will be helpful for other people on this thread.

**Comment by megicivovic** [comment]
@baywet sure thing, I will get back here with some reports.

**Comment by baywet** [comment]
Alright everyone, this issue has been inactive for a while now so I'm going to go ahead and close it.
I believe we tackled most if not all the underlying causes, whether it was in application design, or bugs to fix in the SDK. Thank you all very much for participating, helping to find issues and helping each other!

**Comment by megicivovic** [comment]
@baywet , unfortunately I have to report that the javax.net.ssl.SSLException: Socket closed keeps happening even with 2.6.0 version of Graph client.

**Comment by baywet** [comment]
@megicivovic have you ruled out port exhaustion, memory exhaustion and CPU saturation through your investigation? (in other terms, is the machine and application healthy when this happens)
Do you have data to share around the frequency of that issue? (does it happen all the time, sometimes, how often, how has is evolved those past weeks, is it correlated with specific times of the day and/or operations...)

**Comment by megicivovic** [comment]
According to our monitoring tools, at the time of the latest recording of the exception, all metrics seem in order (CPU and memory way under the limit). As for frequency, we have recorded 172 of these exceptions over last month. It mostly happens for get  calendar events client call. Hope some of this info helps !

**Comment by baywet** [comment]
thanks for this additional information. 172 exceptions over roughly how many requests for the same period? Do you see some pattern like those exceptions happen when the app is making "more" requests to Microsoft Graph?

The service might be intentionally resetting connections to protect itself when it sees to many simultaneous connections from the same origin. This is what I'm trying to rule out at this point. [A good way to mitigate that could be to queue the requests instead of trying to execute them all at once](https://github.com/square/okhttp/issues/1340#issuecomment-77851813). [This is done automatically for you](https://github.com/microsoftgraph/msgraph-sdk-java-core/blob/886811aac3942d88db1213446ff1670fde58da59/src/main/java/com/microsoft/graph/http/CoreHttpProvider.java#L361) when you use our async API that will ship as part of our v3 (currently in preview see #628)

Would it be possible for you upgrade to the preview version and serve 10% of your traffic with this upgraded version to compare the results?

**Comment by megicivovic** [comment]
It was around 1m requests during this period. This is a really small percentage, and I can't seem to make any patterns around it. I am looking forward to trying out your v3. Unfortunately, our infrastructure doesn't support blue/green deployments yet. I think we can hold on until the official release of v3. Thanks for all the support around this issue!

**Comment by abhi-moo** [comment]
Hello, 

I'm using the following dependencies and seeing a lot of exceptions every few minutes in production. Can you please let me know what library this issue is fixed?

    //Microsoft graph API dependencies
    compile group: 'com.microsoft.graph', name: 'microsoft-graph', version: '2.0.0'
    compile(group: 'com.microsoft.graph', name: 'microsoft-graph-auth', version: '0.2.0')

2021-03-23 09:13:35 [scheduling-1] ERROR c.m.e.GraphAPI.service.GraphAPISvc - ERROR Get email failed : Exception: 
com.microsoft.graph.core.ClientException: Error during http request
	at com.microsoft.graph.http.CoreHttpProvider.sendRequestInternal(CoreHttpProvider.java:424)
	at com.microsoft.graph.http.CoreHttpProvider.send(CoreHttpProvider.java:204)
	at com.microsoft.graph.http.CoreHttpProvider.send(CoreHttpProvider.java:184)
	at com.microsoft.graph.http.BaseRequest.send(BaseRequest.java:306)
	at com.microsoft.graph.http.CustomRequest.get(CustomRequest.java:43)
	at com.microsoft.graph.http.CustomRequest$get.call(Unknown Source)
	at com.mutualofomaha.eiw.GraphAPI.service.GraphAPISvc.getEmail(GraphAPISvc.groovy:114)
	at com.mutualofomaha.eiw.GraphAPI.service.GraphAPISvc$getEmail$1.callCurrent(Unknown Source)
	at com.mutualofomaha.eiw.GraphAPI.service.GraphAPISvc.saveMail(GraphAPISvc.groovy:44)
	at com.mutualofomaha.eiw.GraphAPI.service.GraphAPISvc$saveMail$0.call(Unknown Source)
	at com.mutualofomaha.eiw.GraphAPI.batch.EmailBatch.startEiwEmailProcess(EmailBatch.groovy:70)
	at sun.reflect.GeneratedMethodAccessor39.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at org.springframework.scheduling.support.ScheduledMethodRunnable.run(ScheduledMethodRunnable.java:84)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.util.concurrent.FutureTask.runAndReset(Unknown Source)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(Unknown Source)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.lang.Thread.run(Unknown Source)
Caused by: javax.net.ssl.SSLException: Connection reset
	at sun.security.ssl.Alert.createSSLException(Unknown Source)
	at sun.security.ssl.TransportContext.fatal(Unknown Source)
	at sun.security.ssl.TransportContext.fatal(Unknown Source)
	at sun.security.ssl.TransportContext.fatal(Unknown Source)
	at sun.security.ssl.SSLTransport.decode(Unknown Source)
	at sun.security.ssl.SSLSocketImpl.decode(Unknown Source)
	at sun.security.ssl.SSLSocketImpl.readHandshakeRecord(Unknown Source)
	at sun.security.ssl.SSLSocketImpl.startHandshake(Unknown Source)
	at sun.security.ssl.SSLSocketImpl.startHandshake(Unknown Source)
	at okhttp3.internal.connection.RealConnection.connectTls(RealConnection.java:336)
	at okhttp3.internal.connection.RealConnection.establishProtocol(RealConnection.java:300)
	at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:185)
	at okhttp3.internal.connection.ExchangeFinder.findConnection(ExchangeFinder.java:224)
	at okhttp3.internal.connection.ExchangeFinder.findHealthyConnection(ExchangeFinder.java:108)
	at okhttp3.internal.connection.ExchangeFinder.find(ExchangeFinder.java:88)
	at okhttp3.internal.connection.Transmitter.newExchange(Transmitter.java:169)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:41)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:94)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:88)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
	at com.microsoft.graph.httpcore.TelemetryHandler.intercept(TelemetryHandler.java:37)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
	at com.microsoft.graph.httpcore.RedirectHandler.intercept(RedirectHandler.java:123)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
	at com.microsoft.graph.httpcore.RetryHandler.intercept(RetryHandler.java:140)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
	at com.microsoft.graph.httpcore.AuthenticationHandler.intercept(AuthenticationHandler.java:31)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:229)
	at okhttp3.RealCall.execute(RealCall.java:81)
	at com.microsoft.graph.http.CoreHttpProvider.sendRequestInternal(CoreHttpProvider.java:357)
	... 22 common frames omitted
	Suppressed: java.net.SocketException: Connection reset by peer: socket write error
		at java.net.SocketOutputStream.socketWrite0(Native Method)
		at java.net.SocketOutputStream.socketWrite(Unknown Source)
		at java.net.SocketOutputStream.write(Unknown Source)
		at sun.security.ssl.SSLSocketOutputRecord.encodeAlert(Unknown Source)
		... 63 common frames omitted
Caused by: java.net.SocketException: Connection reset
	at java.net.SocketInputStream.read(Unknown Source)
	at java.net.SocketInputStream.read(Unknown Source)
	at sun.security.ssl.SSLSocketInputRecord.read(Unknown Source)
	at sun.security.ssl.SSLSocketInputRecord.readHeader(Unknown Source)
	at sun.security.ssl.SSLSocketInputRecord.decode(Unknown Source)
	... 60 common frames omitted

**Comment by baywet** [comment]
@abhi-moo 
Thanks for reaching out. Your application is behind by a lot, please upgrade to at least:
- microsoft-graph:2.10.0 (3.0 recommended)
- microsoft-graph-core:1.0.9 (2.0.0 recommended)
- auth: 0.3.0 (replacing by azure-identity recommended, see [the upgrade guide](https://github.com/microsoftgraph/msgraph-sdk-java/blob/dev/docs/upgrade-to-v3.md))

**Comment by abhi-moo** [comment]
@baywet 
I've changed the gradle to use the latest as follows:
    //Microsoft graph API dependencies
    compile group: 'com.microsoft.graph', name: 'microsoft-graph', version: '3.0.0'
    // https://mvnrepository.com/artifact/com.azure/azure-identity
    compile group: 'com.azure', name: 'azure-identity', version: '1.2.4'

However, seeing these messages on startup with Azure identity and jackson:
2021-03-23 13:56:48 [scheduling-1] ERROR o.s.s.s.TaskUtils$LoggingErrorHandler - ERROR Unexpected error occurred in scheduled task
java.lang.NoSuchMethodError: com.fasterxml.jackson.databind.ObjectMapper.coercionConfigDefaults()Lcom/fasterxml/jackson/databind/cfg/MutableCoercionConfig;
	at com.azure.core.util.serializer.JacksonAdapter.<init>(JacksonAdapter.java:106)
	at com.azure.core.util.serializer.JacksonAdapter.createDefaultSerializerAdapter(JacksonAdapter.java:139)
	at com.azure.identity.implementation.IdentityClient.<clinit>(IdentityClient.java:94)
	at com.azure.identity.implementation.IdentityClientBuilder.build(IdentityClientBuilder.java:113)
	at com.azure.identity.ClientSecretCredential.<init>(ClientSecretCredential.java:50)
	at com.azure.identity.ClientSecretCredentialBuilder.build(ClientSecretCredentialBuilder.java:62)

2021-03-23 13:56:51 [scheduling-1] ERROR o.s.s.s.TaskUtils$LoggingErrorHandler - ERROR Unexpected error occurred in scheduled task
java.lang.NoClassDefFoundError: Could not initialize class com.azure.identity.implementation.IdentityClient
	at com.azure.identity.implementation.IdentityClientBuilder.build(IdentityClientBuilder.java:113)
	at com.azure.identity.ClientSecretCredential.<init>(ClientSecretCredential.java:50)
	at com.azure.identity.ClientSecretCredentialBuilder.build(ClientSecretCredentialBuilder.java:62)
	at com.azure.identity.ClientSecretCredentialBuilder$build$0.call(Unknown Source)
	at com.mutualofomaha.eiw.GraphAPI.service.AuthSvc.buildGraphClient(AuthSvc.groovy:28)
	at com.mutualofomaha.eiw.GraphAPI.service.AuthSvc$buildGraphClient.call(Unknown Source)

I tried including the latest jackson libraries but did not work.
    //Jackson databind to read yaml
    compile(group: 'com.fasterxml.jackson.dataformat', name: 'jackson-dataformat-yaml', version: '2.11.0')
    compile(group: 'com.fasterxml.jackson.core', name: 'jackson-databind', version: '2.11.0')
    compile(group: 'com.fasterxml.jackson.core', name: 'jackson-annotations', version: '2.11.0')

Thanks for helping, looking forward to hear back and resolve this.

**Comment by abhi-moo** [comment]
I was able to get passed the above exception by using the following version:
    compile(group: 'com.fasterxml.jackson.dataformat', name: 'jackson-dataformat-yaml', version: '2.12.0')
    compile(group: 'com.fasterxml.jackson.core', name: 'jackson-databind', version: '2.12.0')
    compile(group: 'com.fasterxml.jackson.core', name: 'jackson-annotations', version: '2.12.0')
    compile(group: 'com.fasterxml.jackson.core', name: 'jackson-core', version: '2.12.0')

I'm using the same credentail secret used which was set in Azure AD application. FYI, this is a windows service backend app reading emails and doing some process -> eventually forwarding, deleting the email.

**Now, I'm seeing this Auth token exception:**
2021-03-23 14:15:15 [ForkJoinPool.commonPool-worker-9] ERROR c.m.a.m.ConfidentialClientApplication - ERROR [Correlation ID: 8198e24a-936e-4965-9297-dfcfc4daa074] Execution of class com.microsoft.aad.msal4j.AcquireTokenByAuthorizationGrantSupplier failed.
com.microsoft.aad.msal4j.MsalServiceException: AADSTS7000215: Invalid client secret is provided.
Trace ID: f176f0f7-61e2-4a07-9c14-7b2090b3cc00
Correlation ID: 8198e24a-936e-4965-9297-dfcfc4daa074
Timestamp: 2021-03-23 19:15:15Z
	at com.microsoft.aad.msal4j.MsalServiceExceptionFactory.fromHttpResponse(MsalServiceExceptionFactory.java:43)
	at com.microsoft.aad.msal4j.TokenRequestExecutor.createAuthenticationResultFromOauthHttpResponse(TokenRequestExecutor.java:96)
	at com.microsoft.aad.msal4j.TokenRequestExecutor.executeTokenRequest(TokenRequestExecutor.java:37)
	at com.microsoft.aad.msal4j.AbstractClientApplicationBase.acquireTokenCommon(AbstractClientApplicationBase.java:120)
	at com.microsoft.aad.msal4j.AcquireTokenByAuthorizationGrantSupplier.execute(AcquireTokenByAuthorizationGrantSupplier.java:63)
	at com.microsoft.aad.msal4j.AuthenticationResultSupplier.get(AuthenticationResultSupplier.java:59)
	at com.microsoft.aad.msal4j.AuthenticationResultSupplier.get(AuthenticationResultSupplier.java:17)
	at java.util.concurrent.CompletableFuture$AsyncSupply.run$$$capture(CompletableFuture.java:1590)
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java)
	at java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1582)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
2021-03-23 14:15:15 [ForkJoinPool.commonPool-worker-9] ERROR c.a.identity.ClientSecretCredential - ERROR Azure Identity => ERROR in getToken() call for scopes [https://graph.microsoft.com/.default]: AADSTS7000215: Invalid client secret is provided.
Trace ID: f176f0f7-61e2-4a07-9c14-7b2090b3cc00
Correlation ID: 8198e24a-936e-4965-9297-dfcfc4daa074
Timestamp: 2021-03-23 19:15:15Z

**Comment by abhi-moo** [comment]
@baywet 
The above issue is resolved. I forgot to decrypt the secret key. I'll leave the above post as it may help others.

Thanks for now.

**Comment by abhi-moo** [comment]
@baywet 
I need your recommendation about instantiating GraphServiceClient graphClient. 
Our background service using GraphAPI does these following:

1. Read 50 emails per batch run. Instantiate new GraphServiceClient as:
```
                //Get all the email messages from the specified mailbox
                MessageCollectionPage messages = eiwAuthSvc.buildGraphClient().users(mailBox.getSource()).mailFolders(INBOX).messages().buildRequest(eiwUtil.getTextBodyType()).top(50).get()
```

2. Forward the email to backup mailbox:
```
            eiwAuthSvc.buildGraphClient().users(emailBox.getSource()).messages(mailId).forward(MessageForwardParameterSet
                    .newBuilder()
                    .withToRecipients(Arrays.asList(recipient))
                    .build()
            ).buildRequest().post()
```

3. Process email - Read email, process attachments and writing content to a network share
` bufferedInputStream = eiwAuthSvc.buildGraphClient().users(email).messages(mailId).content().buildRequest(options).get()
`
4. Delete the email
`            eiwAuthSvc.buildGraphClient().users(emailBox.getSource()).messages(mailId).buildRequest(options).delete()
`

**Question:** 
1) Should I instantiate new graphClient everytime OR can we instantiate graphClient once and pass it?
2) What is the max timeout for graphClient? Will it be alive till the batch process terminates?

**Comment by baywet** [comment]
@abhi-moo your last few comments are digressing from the original discussion point, and there's a lot of people on this thread.
In a effort to keep this discussion focused and save people from notifications, would you mind opening a new issue please? Thanks!

**Comment by sujitjadhav07** [comment]
@baywet I have upgraded my application (uses spring boot) to microsoft-graph 3.0 and followed the upgrade guide.
However im getting the socket errors after upgrading. More details on stackoverflow [here](https://stackoverflow.com/questions/66820194/microsoft-graph-3-0-java-net-sockettimeoutexception-while-retrieving-group-membe)
